0000-overhaul-qiskit-experiments.md

```diff
@@ -0,0 +1,638 @@
+exc.run()
+
+exc.experiment_data.block_for_results()
+exc.experiment_data.save()
+```
+
+Note that [`BaseAnalysis.run`](https://github.com/Qiskit-Extensions/qiskit-experiments/blob/c66034c90dad73d705af25be7e9ed9617e7eb2ef/qiskit_experiments/framework/base_analysis.py#L118-L123) is not called in the new workflow.
+This method conventionally takes `ExperimentData` and mutates the input container through a callback.
+It internally wraps the `_run_analysis` method as a local function in which created analysis data are formatted and added to the experiment data.
+This local function is added to the input experiment data as a callback (very complicated!).
+By design, the analysis callback must NOT mutate the experiment data because the callback can run on a subprocess,
+and the mutated experiment data container doesn't update the original container residing in the main process.
+Therefore, the `BaseAnalysis.run` method is going to be deprecated, and instead we can add
+
+```python
+class BaseAnalysis:
+    ...
+
+    def __call__(
+        self, 
+        experiment_data: ExperimentData,
+    ) -> List[AnalysisResultData, Figure]:
+        return self._run_analysis(experiment_data)
+```
+
+to turn analysis instance into callable. This allows the executor to directly submit analysis instances.
+Eventually, we can also deprecate the `_run_analysis` method.
+Experiment developer will directly overwrite the call dunder method to provide the implementation of analysis.
+
+## Alternative Approaches
+In the proposed design, the executor is hidden and the experiment class behaves as the sole user interface.
+Alternatively, we can expose executor and remove `BaseExperiment.run` method.
+
+```python
+with Executor() as exc:
+    exp = MyExperiment((0,), **options)
+    exp_data = exc.run(exp)
+```
+
+On the one hand, this syntax has a high affinity with Runtime session, and we can completely delegate
+the role of circuit and analysis execution to the `Executor`. Experiment class becomes just a factory of experimental circuits.
+On the other hand, this introduces a breaking API change also to end users, and preclude us from capability of single liner execution with method chain.
+
+```python
+exp_data = MyExperiment(
+    (0,), backend=backend, **options
+).run().block_for_results()
+```
+
+If we have enough bandwidth to support end-user migration, alternative implementation is worth considering.
+
+
+## Questions
+How to communicate with community developers?
+How to avoid breaking API change?
+What is the reasonable timeline for migration?
+
+Another important situation that we may need to cover is the break of the executor run.
+Job queueing in a quantum system may take very long time (sometime more than a day) if you don't reserve the device, 
+and we often terminate the kernel where the circuit executor is actively running.
+In this situation, we may want to restart from the analysis, once after all jobs successfully complete.
```

Comment:
I think we should also add an option to not run the analysis automatically after job completion, like @TsafrirA's use case.

Reply:
Yes. We can continue to support `Experiment.run(analysis=None)`. In this case the analysis executor will receive an empty task dependency graph and run nothing.